packages/stream-tests/src/__tests__/basic-indexing.test.ts

```diff
@@ -0,0 +1,252 @@
+  })
+
+  test('multiple documents - one page, reverse order', async () => {
+    const doc1 = await ModelInstanceDocument.create(ceramic, CONTENT0, midMetadata)
+    await doc1.replace(CONTENT1)
+    const doc2 = await ModelInstanceDocument.create(ceramic, CONTENT2, midMetadata)
+    const doc3 = await ModelInstanceDocument.create(ceramic, CONTENT3, midMetadata)
+
+    console.log(`docIds: [${doc1.id.toString()}, ${doc2.id.toString()}, ${doc3.id.toString()}]`)
+
+    const resultObj = await ceramic.index.queryIndex({ model: model.id, last: 100 })
+    const results = await extractDocuments(ceramic, resultObj)
+
+    expect(results.length).toEqual(3)
+    expect(results[0].id.toString()).toEqual(doc3.id.toString())
+    expect(results[0].content).toEqual(CONTENT3)
+    expect(results[1].id.toString()).toEqual(doc2.id.toString())
+    expect(results[1].content).toEqual(CONTENT2)
+    expect(results[2].id.toString()).toEqual(doc1.id.toString())
+    expect(results[2].content).toEqual(CONTENT1)
+  })
+
+  test('multiple documents - multiple pages, reverse order', async () => {
+    const doc1 = await ModelInstanceDocument.create(ceramic, CONTENT0, midMetadata)
+    await doc1.replace(CONTENT1)
+    const doc2 = await ModelInstanceDocument.create(ceramic, CONTENT2, midMetadata)
+    const doc3 = await ModelInstanceDocument.create(ceramic, CONTENT3, midMetadata)
+    const doc4 = await ModelInstanceDocument.create(ceramic, CONTENT4, midMetadata)
+    const doc5 = await ModelInstanceDocument.create(ceramic, CONTENT5, midMetadata)
+
+    const resultObj0 = await ceramic.index.queryIndex({ model: model.id, last: 2 })
+    expect(resultObj0.pageInfo.hasPreviousPage).toBeTruthy()
+    const resultObj1 = await ceramic.index.queryIndex({
+      model: model.id,
+      last: 2,
+      before: resultObj0.pageInfo.endCursor,
+    })
+    expect(resultObj1.pageInfo.hasPreviousPage).toBeTruthy()
+    const resultObj2 = await ceramic.index.queryIndex({
+      model: model.id,
+      last: 2,
+      before: resultObj1.pageInfo.endCursor,
+    })
+    expect(resultObj2.pageInfo.hasPreviousPage).toBeFalsy()
+
+    const results = (
+      await Promise.all([
+        extractDocuments(ceramic, resultObj0),
+        extractDocuments(ceramic, resultObj1),
+        extractDocuments(ceramic, resultObj2),
+      ])
+    ).flat()
+
+    expect(results.length).toEqual(5)
+    expect(results[0].id.toString()).toEqual(doc5.id.toString())
+    expect(results[0].content).toEqual(CONTENT5)
+    expect(results[1].id.toString()).toEqual(doc4.id.toString())
+    expect(results[1].content).toEqual(CONTENT4)
+    expect(results[2].id.toString()).toEqual(doc3.id.toString())
+    expect(results[2].content).toEqual(CONTENT3)
+    expect(results[3].id.toString()).toEqual(doc2.id.toString())
+    expect(results[3].content).toEqual(CONTENT2)
+    expect(results[4].id.toString()).toEqual(doc1.id.toString())
+    expect(results[4].content).toEqual(CONTENT1)
```

Comment:
Looks like the order of the results is reversed here.
My understanding from the connection spec is that if we have `[A, B, C, D, E, F]` and we query `last: 2`, we should get `[E, F]` with `{ hasNextPage: false, hasPreviousPage: true, startCursor: E, endCursor: F }`, then `last: 3, before: E` should return `[B, C, D]`.
Essentially when using `last` we iterate backwards but that shouldn't change how the slices of results are returned.

Reply:
woah.  That makes no sense to me at all.  What if you just want to iterate in literal reverse order?  Is there no way to do that in GraphQL?

Reply:
isn't that also going to be way less efficient in the database as you can't just iterate the index in reverse order?

Reply:
It looks like GraphQL is supposed to also support an `order` argument: https://dgraph.io/docs/graphql/queries/order-page/.  We don't currently seem to support that - should we?  Being able to query in reverse order seems way more useful to me than being able to fetch the `last` N items in forward order would be.

Reply:
I guess we removed that when we stopped supporting multiple orderings.  I didn't realize at the time that by removing support for ordering on different groups of fields we also removed support for querying in reverse order :disappointed: 

Probably too late to add to the upcoming developer preview release now, but seems like it should be added as part of one of the follow-ups?  Either the one with relations or the one with unions and interfaces.  What do you think @PaulLeCam?

Reply:
> It looks like GraphQL is supposed to also support an `order` argument: https://dgraph.io/docs/graphql/queries/order-page/. We don't currently seem to support that - should we? Being able to query in reverse order seems way more useful to me than being able to fetch the `last` N items in forward order would be.

This is not the official GraphQL spec, seems more like a custom implementation on top with its own additions, no?

The official spec isn't opinionated about custom ordering, we can do what we want for this, but what matters is that the pagination slices are consistent, cf https://relay.dev/graphql/connections.htm#sec-Edge-order

> When before: cursor is used, the edge closest to cursor must come last in the result edges.
> When after: cursor is used, the edge closest to cursor must come first in the result edges.
 
That seems pretty logical to me, no? It's mainly a convention so the GraphQL servers and clients don't have to worry about different pagination mechanisms I guess.



> I guess we removed that when we stopped supporting multiple orderings. I didn't realize at the time that by removing support for ordering on different groups of fields we also removed support for querying in reverse order 😞

Not sure I follow, what is the problem with querying in reverse order? GraphQL connections support using either `first` with optional `after` or `last` with optional `before` arguments, so the ordering can be applied as requested by the client, no?

Reply:
>  what is the problem with querying in reverse order?

So let's say you create streams A, B, C, D, and E in order.  A normal forward query will return ABCDE.
If you use `last` with a page size of 2 you get back DEBCA (or if you use `last` with a large enough page size, you just get back ABCDE again).  That doesn't seem like a very useful ordering to me.  It seems like we should have a way for users to get back EDCBA - actual reverse order

Reply:
You don't actually get DEBCA though, you first get [DE], then [BC] that you need to prepend to the results list, so now you have [BCDE] and then you get [A] that you prepend again, and get [ABCDE].
In the client you can reverse the result list if needed, or simply iterate from the end to the start of the list, that's not really a spec problem at this point.

I guess one thing to remember here is that this is designed for cursor-based access, the ordering of queries is a "just in time" concern.
Say you want to query messages in a live chat, you start by querying the last 2 messages for example, you get [EF]. Now you also start polling for new messages, start getting [G], then [H] that you just append, so you have [EFGH]. The user scrolls up the message lists, so we want to load the last 2 messages messages before E, we get back [CD] that needs to be prepended to the list, which becomes [CDEFGH] and is likely what the app wants, no?

Reply:
This sounds like the app is managing all these pages manually.  I had thought the client would present a simple iterator interface for the client where they could just get results one by one with a `next()/hasNext()` API, and all the pagination would be handled automatically by the client.  Is the client really expected to understand all the nuances around startCursor vs endCursor, before vs after, last vs first etc?  I only just got my head around it - is it really the expectation that every app dev will need to do the same?

Reply:
Depends how high-level the GraphQL client goes, but for example with Relay this is handled as you describe: https://relay.dev/docs/guided-tour/list-data/pagination/